java/src/main/java/io/cucumber/java/JavaHookDefinition.java

```diff
@@ -22,41 +29,88 @@
 
     private static Method requireValidMethod(Method method) {
         Class<?>[] parameterTypes = method.getParameterTypes();
-        if (parameterTypes.length > 1) {
-            throw createInvalidSignatureException(method);
+        if (parameterTypes.length > 0) {
+            Class<?>[] annotationTypes = getRelevantMethodAnnotations(method.getAnnotations());
+            for (Class<?> parameterType : parameterTypes) {
+                checkParameter(method, parameterType, annotationTypes);
+            }
         }
+        return method;
+    }
 
-        if (parameterTypes.length == 1) {
-            Class<?> parameterType = parameterTypes[0];
-            if (!(Object.class.equals(parameterType) || io.cucumber.java.Scenario.class.equals(parameterType))) {
-                throw createInvalidSignatureException(method);
+    private static Class<?>[] getRelevantMethodAnnotations(Annotation[] annotations) {
+        if (null == annotations || annotations.length == 0) {
+            return new Class<?>[0];
+        }
+        Class<?>[] relevant = null;
+        for (Annotation annotation : annotations) {
+            Class<? extends Annotation> annotationType = annotation.annotationType();
+            if (Before.class.equals(annotationType) || After.class.equals(annotationType)) {
+                return new Class<?>[] {Before.class, After.class};
+            }
+            if (BeforeStep.class.equals(annotationType) || AfterStep.class.equals(annotationType)) {
+                relevant = new Class<?>[] {BeforeStep.class, AfterStep.class};
             }
         }
+        return relevant;
+    }
 
-        return method;
+    private static Set<Class<?>> getAcceptedTypes(Class<?> annotationType) {
```

Comment:
This method changes the behaviour of the `JavaHookDefinition` based on what it is annotated with. This indicates that `After` and `Before` annotated steps are different from `BeforeStep` and `AfterStep` annotated classes. When there is a difference in behaviour it makes sense to create a separate class for each. This should reduce the complexity of this class significantly.